프로젝트 매니저 [ STEP2 ] 웡빙

wongbingg opened the pull request:
안녕하세요 제임스 ! @inwoodev  step2 수행하여 PR 보냅니다! 
디자인 패턴에 대한 적용이 완벽하게 되지 않은 상황입니다..! 제가 이해한 부분까지 적용을 해보았는데 수정 해나갈 방향에 대해 조언을 듣고 싶습니다 🥲🥲 잘 부탁 드립니다!! 🫠🫠

# 📁 Project Manager

###### tags: `README`

## 개요

할 일 목록을 **TODO**, **DOING**, **DONE** 세 가지 카테고리로 나누어 유연하게 관리해주는 앱.

## 기술스택
| Dependency | UI | Design Pattern |
|:---:|:---:|:---:|
|Realm, FireBase, SwiftLint|UIKit|MVVM + Coordinator|

## File Tree
```
.
├── ProjectManager
│   ├── swiftlint
│   ├── ProjectManager
│   │   ├── Protocol
│   │   │   └── Coordinator
│   │   ├── Extensions
│   │   │   ├── UITextField+Extension
│   │   │   └── TimeInterval+Extension
│   │   ├── Application
│   │   │   ├── AppDelegate
│   │   │   ├── AppAppearance
│   │   │   └── AppCoordinator
│   │   ├── MainScene
│   │   │   ├── ViewModel
│   │   │   │   └── TodoListViewModel
│   │   │   └── View
│   │   │       ├── TodoListViewController
│   │   │       └── ListView
│   │   │           ├── Header
│   │   │           │   └── HeaderView
│   │   │           └── CollectionView
│   │   │               ├── ListCollectionView
│   │   │               └── Cell
│   │   │                   └── ListCell
│   │   ├── ModalTransitionScene
│   │   │   └── View
│   │   │       ├── CreateTodoListViewController
│   │   │       ├── EditTodoListViewController
│   │   │       └── Template
│   │   │           └── FormSheetTemplateView
│   │   ├── PopoverScene
│   │   │   ├── ViewModel
│   │   │   │   └── PopoverViewModel
│   │   │   └── View
│   │   │       └── PopoverViewController
│   │   ├── Model
│   │   │   ├── TodoModel
│   │   │   └── Category
│   │   ├── mockData
│   │   ├── Assets
│   │   ├── LaunchScreen
│   │   └── Info
│   ├── Products
│   ├── Pods
│   └── Frameworks
└── Pods
```
## 화면구성
![](https://i.imgur.com/aYpgYgB.png)

## 기능구현

### - AppCoordinator
- 뷰 전환을 AppCoordinator에서 관리 하도록 구현
- 각 ViewController 에 
```swift
static func create(with: ViewModel, coordinator: Coordinator) -> UIViewController
``` 
위와 같은 형태의 메서드를 이용하여 ViewController를 생성할 때, ViewModel과 Coordinator 를 주입

### - 할 일 추가
- "TodoListViewController" 에서 네비게이션 바의 우측상단 "+" 버튼을 누르면 modalTransitionStyle 이 .formSheet 형태로 입력양식을 띄움
    - 입력 양식에서 우측상단의 Done 버튼 탭시, 입력된 데이터를 저장 후 셀 생성
    - 입력 양식에서 좌측상단의 Cancel 버튼 탭시, 취소
- 이 루트를 통해 생성되는 셀은 모두 TODO List 로 추가 되도록 설정

<img src="https://i.imgur.com/Yu97jAJ.gif" width=500>

### - 할 일 삭제
- 해당 셀 trailing Swipe시, 삭제버튼이 나오도록 구현
    - 나타난 delete 버튼 탭시 셀 삭제
- 해당 셀 trailing Swipe를 끝까지할 시, 삭제
    
<img src="https://i.imgur.com/3ijJ8ux.gif" width=500>

### - 할 일 수정
- 특정 셀 탭시, modalTransitionStyle이 .formSheet 형태로 입력양식을 띄움
    - 입력 양식에는 해당 셀의 세부정보가 채워진다
    - 입력 양식에서 내용을 수정한 뒤, 좌측상단의 Edit 버튼 탭시 셀 정보 수정
    - 입력 양식에서 Done 버튼 탭시, 뒤로가기 (수정사항 반영 x)

<img src="https://i.imgur.com/vcFUzng.gif" width=500>

### - 할 일 이동
- 특정 리스트 내에서 LongTouch를 할 시, 나머지 리스트로 이동할 수 있는 modalTransitionStyle이 .popover인 양식 띄움.
- 두 버튼에 어느 리스트로 이동 할 것인지 표현, 버튼 탭시 해당 리스트로 셀을 이동.

<img src="https://i.imgur.com/58NYrYN.gif" width=500>

### - 기타
- HeaderView
    - 해당 리스트의 현재 셀 개수를 HeaderView.count에 표현
- ListCell
    - 날짜가 지나면 dateLabel의 색을 빨간색으로 변경
    - 제목이 길면 끝에 ...으로 표현
    - 본문은 최대 3줄까지만 표현

## ⚒🛠Trouble Shooting

### 1. ViewModel 에서 UIKit 제거하기

ViewModel의 view transition 파트에서 파라미터로 CGPoint를 받아오는 부분 때문에 import UIKit이 되어있었다. 불필요한 부분이라 상위함수에서 전달해줄 때, CGPoint의 형태를 (Double, Double) 튜플형태로 전달 해주었다.
[근거](https://dev200ok.blogspot.com/2021/07/006-float-double-cgfloat.html)
> 추가적으로 Swift 5.5 버젼 이상에서는 CGFloat과 Double이 교환 가능하다고 합니다.

## 질문사항 

### 1. 네비게이션 바 구성

NavigationContorller 관련 코드를 모두 Coordinator로 빼주려 했지만, 불가능한 부분이 있었습니다. Coordinator를 사용하면 navigationBar 설정 (ex: 네비게이션 타이틀, 바 버튼아이템) 또한 Coordinator에서 설정 해주는게 맞을까요 ?

### 2. 뷰 모델 구성

TodoListViewModel 하나를 대부분의 뷰에서 사용하고 있는데, 각 뷰마다의 뷰모델이 존재해야 바람직한 것인지 궁금합니다! 

### 3. 디자인 패턴

MVVM 패턴의 구조를 완벽히 이해하지 못하여 흉내내기 식으로 구현을 해봤는데, 이런 구현이 맞는지가 궁금합니다..! 
또한 MVVM 예시코드 에서는 모두 뷰 모델을 프로토콜로 추상화를 시켜주었던데, 이 부분처럼 저도 수정해 나갈 수 있을지 궁금합니다..! 그리고 테스트코드를 짜고 싶었는데 아직 이부분도 어떻게 해야할지 감이 안와 수행하지 못했습니다...ㅜㅜ 전반적인 설계방향에 대한 조언을 듣고 싶습니다!

## 참고 링크
[Apple Article: Displaying transient content in a popover](https://developer.apple.com/documentation/uikit/windows_and_screens/displaying_transient_content_in_a_popover)
[LongTouchGestureRecognizer사용법](http://yoonbumtae.com/?p=4418)
[스크롤 시, 네비게이션 바 자동 숨김처리 비활성화](https://nemecek.be/blog/126/how-to-disable-automatic-transparent-navbar-in-ios-15)
[UIView그림자 만들기](https://babbab2.tistory.com/41)

wongbingg committed:
feat : NavigationBar 설정

- 스크롤시 네비게이션바가 사라지지 않도록 appearance 설정
- 네비게이션바 title 과 rightBarButtonItem 구성

wongbingg committed:
feat: coordinator 세팅

- SceneDelegate 제거
- AppCoordinator 구현
- AppDelegate 내 AppCoordinator 적용

wongbingg committed:
feat : CollectionView 생성

- 재사용을 위한 CollectionView

wongbingg committed:
feat : CollectionView + HeaderView = ListView

- 최종 재사용 단위 ListView 생성

wongbingg committed:
feat : MainScene 생성

1. View : "TodoListViewController"
    - View 내에 3개의 ListView 를 주입
    - AppCoordinator 에서 ViewModel을 주입 후 생성
2, ViewModel: "TodoListViewModel"
    - 데이터에 대한 CRUD 구현
    - 클로져를 활용한 데이터 바인딩
    - coordinator 를 활용한 화면전환 구현

wongbingg committed:
feat : FormSheet 생성

- "FormSheetTemplateView" 를 이용하여 "Create", "Edit" FormSheet ViewController 생성
- "Create" 와 "Edit" 버튼액션 수행 시 주입된 viewModel의 메서드 이용

wongbingg committed:
feat : "PopoverViewController" 생성

- 셀 길게 터치 시 나타나는 PopoverViewController 생성
- PopoverViewController의 버튼액션을 위해 "PopoverViewModel" 생성
- PopoverViewModel에서 버튼액션 수행을 위해 상위 뷰모델인 "TodoListViewModel"을 사용

wongbingg committed:
feat : UITextField 와 TimeInterval 의 Extension 구현

- etc : 임시 mockData 생성

wongbingg committed:
feat : RealmSwift 라이브러리 추가

- cocoaPod을 이용하여 의존성 추가

wongbingg committed:
style: 컨벤션 통일

- 마크주석으로 코드의 기능분리
- 불필요한 개행 제거
- 불필요한 주석 제거
- 프로퍼티 선언 순서를 접근제어자 순서로 지정

wongbingg committed:
refactor : 들여쓰기 코드 수정, ViewModel의 Import UIKit 제거

- etc : "FormSheetTemplateView" 의 setupData(with:) 메서드 수정

wongbingg committed:
docs : ReadMe 업데이트

inwoodev commented on ProjectManager/Podfile.lock:
```diff
@@ -1,16 +1,26 @@
 
-PODFILE CHECKSUM: a184e704be7b82ce48a8d528fbd0d049f65235b5
+PODFILE CHECKSUM: e4ea0896360b6539857b66d1cd5699b9e6e0317c
 
```
podfile.lock에 대해서 공부해 보셨는지 궁금합니다~(진짜로 궁금해서 적었어요 ㅎㅎ)

wongbingg replied:
podfile.lock은 현재  제가 받은 의존성인 `SwiftLint`와 `RealmSwift` 의 버전이 입력 되어 다른 누군가가 클론을 받았을 때 해당 podfile을  이 podfile.lock 에 명시된 버전으로 pod install 받을 수 있도록 pod들의 버전을 추적하여 기록하고 유지시켜주는 역할을 한다고 알고있습니다. 😀

inwoodev commented on ProjectManager/ProjectManager/MainScene/View/TodoListViewController.swift:
```diff
@@ -0,0 +1,165 @@
+        bind()
+    }
+    
+    static func create(with viewModel: TodoListViewModel,
+                       coordinator: AppCoordinator) -> TodoListViewController {
+        let viewController = TodoListViewController()
+        viewModel.coordinator = coordinator
+        viewController.viewModel = viewModel
+        return viewController
+    }
```
음 이 메서드는모든 뷰컨트롤러에 해당 메서드가 존재하는거 같네요~ extension으로 빼고 프로토콜, generic등을 적절히 사용한다면 더 범용적이게 사용할 수 있을거 같은데요? 어떻게 생각하시나요? :)

wongbingg replied:
좋은 생각인 것 같습니다! ㅎㅎ 하지만 추상화에 대한 개념이 아직 제대로 잡혀있지 않아 이 부분을 조금 더 공부해본 후 수정해 보겠습니다! 수정중 모르는 점이 생기면 질문드리겠습니다 ㅎㅎ

inwoodev commented on ProjectManager/ProjectManager/MainScene/View/TodoListViewController.swift:
```diff
@@ -0,0 +1,165 @@
+    
+    private func didCreatedItem() {
+        viewModel?.didCreatedItem = { [weak self] (item) in
+            guard let self = self else { return }
+            switch item.category {
```
`guard let self = self else { return}` 구문은 온라인상에서 한 때 많이 논의되었던 구문인걸로 기억해요. 혹시 조금 찾아보시고 공부해 보신 내용을  공유받을 수 있을까요?

wongbingg replied:
[Besher Al Maleh Articel](https://medium.com/@almalehdev/you-dont-always-need-weak-self-a778bec505ef)

위 아티클 내용을 확인 해보았습니다 ! 완벽하게 파악하지는 못했지만 [weak self] 자체도 무조건 필요한 게 아닌, Non-escaping closure의 경우엔 요구되지 않는다는 것을 알았습니다. 경우에 따라 적절히 사용해야겠군요..ㅎㅎ
또한 guard let 구문으로 인해 closure 내부에 있는 동안 일시적인 강한참조를 하여 self를 사용할 수 있다고 내용이 나와 있었습니다. 이로 인해 delayed deallocate 라는 현상이 발생한다고 합니다. 
 self?. 로 접근하는 것과 guard let 구문의 차이점으로 만약 클로저 실행 중간에 화면이 dismiss 되어 self(=viewController) 가 nil이 되었을 시 self?. 으로 접근하는 경우 메서드들은 skip이 될 것이고,  guard let 으로 일시적인 강한참조를 시켜놓은 self(=viewController) 의 경우 nil이 되지 않고 할당해제가 지연 될 것입니다. 
 
 결론적으로, ViewController가 해제된 이후에 불필요한 작업진행이 되는 것을 피하고 싶다면 self?. 으로 접근을 하고
 객체가 해제되기 전에 모든 작업을 완료하고 싶다면 guard let 구문을 이용해라 ! 라는 내용으로 파악을 했습니다.ㅎㅎ

inwoodev commented on ProjectManager/ProjectManager/MainScene/View/ListView/CollectionView/ListCollectionView.swift:
```diff
@@ -0,0 +1,189 @@
+}
+
+// MARK: - CollectionView Delegate
+extension ListCollectionView: UICollectionViewDelegate {
```
listCollectionview의 로직을 대신하기 위한 대리자 역할을 자기 자신이 하고 있네요. 그리고 밑에 있는 `UIGestureRecognizerDelegate` 또한 collectionView가 채택하고 있어요. 음 mvvm과 mvc과 같은 디자인패턴이 나온 이유는 여러가지 있겠지만 그 중 가장 큰 이유는 뷰에서 비즈니스 로직을 분리하기 위해서라고 생각하는데요. 이와 같은 구조로 계속 설계하게 된다면 추후 확장과 분리에 있어서 어려움을 겪지 않을까요?

wongbingg replied:
뷰에서 비즈니스 로직을 분리하는 작업이 아직 어떤 것인지, 어떤 기준인지 제대로 파악하지 못한 것 같습니다 😂 일단 collectionViewDelegate 를 TodoListViewController에서 채택을 하도록 수정을 해주었는데 제가 코멘트의 의도를 잘 파악한 것인지 모르겠습니다😂 이 부분 자세하게 설명을 요청드려도 될까요 ?

inwoodev reviewed (commented):
와우 전반적으로 코드가 매우 깔끔해서 리뷰 하기 너무 수월했어요 :) 고생하셨습니다~ 리뷰하면서 저도 많이 자극 받았어요🤩 웡빙 덕분에 조금 잦아들었던 개발의 열정이 더 불타오르게 된거 같아요 ㅎㅎ. 리뷰하면서 느낀점을 코멘트로 남겨봤습니다. 확인 해 주시고 질문이 있으시면 언제든 comment 또는 dm 주세요~

> 1. 네비게이션 바 구성
NavigationContorller 관련 코드를 모두 Coordinator로 빼주려 했지만, 불가능한 부분이 있었습니다. Coordinator를 사용하면 navigationBar 설정 (ex: 네비게이션 타이틀, 바 버튼아이템) 또한 Coordinator에서 설정 해주는게 맞을까요 ?

음 이건 제 생각이에요:) 참고만 해주세요:) 저 같은 경우 coordinator의 역할은 화면전환에만 집중하는 게 더 깔끔하다고 생각합니다. 솔리드 원칙에서도 단일책임 원칙이 나오잖아요~ 화면전환에만 집중해서 설계를 하게 되면 더 재사용성이 높은 코드를 작성할 수 있다고 생각하구요 coordinator또한 같은 맥락에서 드리는 말씀입니다~ 아참 그리고 이번 프로젝트에서 coordinator가 navigation controller를 가지고 있어야 할지도 생각해보면 좋을거 같아요!

---
> 2. 뷰 모델 구성
TodoListViewModel 하나를 대부분의 뷰에서 사용하고 있는데, 각 뷰마다의 뷰모델이 존재해야 바람직한 것인지 궁금합니다!

뷰모델이라는 개념에 대해서 조금 더 같이 고민 해 봐요 우리 ㅎㅎ. 제 생각에는 뷰모델이라는 개념이 생긴 가장 큰 이유 중 하나는 매시브한 뷰컨트롤러 때문에 유지보수하기가 어려워지기 때문이겠죠? 이런 관점에서 본다면 하나의 뷰모델로 모든 비즈니스로직을 관리하기가 어렵다면 필요에 따라 특정뷰만을 위해서 추가로 뷰모델을 설계해 볼 수 있을거 같아요

---
> 3. 디자인 패턴
MVVM 패턴의 구조를 완벽히 이해하지 못하여 흉내내기 식으로 구현을 해봤는데, 이런 구현이 맞는지가 궁금합니다..!
또한 MVVM 예시코드 에서는 모두 뷰 모델을 프로토콜로 추상화를 시켜주었던데, 이 부분처럼 저도 수정해 나갈 수 있을지 궁금합니다..! 그리고 테스트코드를 짜고 싶었는데 아직 이부분도 어떻게 해야할지 감이 안와 수행하지 못했습니다...ㅜㅜ 전반적인 설계방향에 대한 조언을 듣고 싶습니다!

먼저 언급하고 싶은 건 제 생각에는 mvvm패턴은 개발의 답이 아니라는거입니다. mvvm패턴을 활용했는데 boilder plate코드가 오히려 더 많아질 수도 있지 않을까요? mvvm패턴의 단점도 분명히 존재합니다. 그럼에도 불구하고 현업에서 mvvm패턴을 많이들 지향하는 이유는 mvvm패턴이 제공하는 장점이 단점을 상쇄하기 때문이지 않을까요? mvvm패턴의 장점을 찾아보시고 그 장점을 극대화 시킬 수 있는 방향으로 개발 해 보시는건 어떨까요 :)

제 생각에 mvvm의 장점은 다음과 같습니다:
1. 반응형이라는 점
2. 뷰는 뷰모델의 존재를 인지하지만 뷰모델은 뷰에 대해서 몰라도 되기 때문에 ViewModel의 객체는 UI와 전혀 상관없는 독립적인 단위 클래스가 되고 이때문에 테스트하기 용이해진다는 점
3. 두 번째에서 이어지는 부분이긴 한데요. input과 output을 모델을 활용한다면 단방향으로 데이터 흐름을 전개하는 게 가능해지고 이 때문에 로직의 흐름을 파악하기 쉽고 테스트하기 수월해진다는 점

위에 언급한 부분을 조금 더 파보시면 mvvm과 더 친해지시지 않을까 생각합니다 :) 그리고 웡빙이 다른 장점들 또는 단점들을 찾으시고 공유 해 주시면 같이 얘기 해 보면서 같이 더 성장할 수 있겠네요~

그리고 질문 중에 부모델을 프로토콜로 추상화 시킨 예시를 많이 보셨다고 하셨는데요. 그 이유도 혹시 찾아보셨는지 궁금합니다. 추상화와 유닛테스트와 관계가 있을까요? 조금 더 고민 해 보시면 좋겠습니다~

---

몇 자 남겨봤어요. 제가 남긴 답변에서 틀린점이 있을 수도 있겠죠? 틀린점 발견해 주시면 언제든 지적해주세요~~ 파이팅입니다!!

wongbingg committed:
refactor : 불필요한 파일 제거

wongbingg committed:
refactor : Coordinator 코드 수정

- 각 ViewController에서 Coordinator를 가지도록 수정

wongbingg committed:
feat : Realm을 이용해 TodoDataManager 생성

- Todo 모델타입 생성
- TodoModel타입을 이용했던 곳을 Todo 타입으로 변경
- 기존의 TodoModel 삭제

wongbingg committed:
refactor : ListCollectionView 수정

- UICollectionViewDelegate Extension을 TodoListViewController로 이동
- GestureRecognizerDelegate Extension을 파일분리
- UIPopoverPresentationControllerDelegate Extension을 파일분리
- TodoListCoordinator에 popoverViewController로의 뷰전환 시점전달을 위해 transitionDelegate 생성

wongbingg committed:
refactor : ViewController 생성 메서드 제거

- static func 대신, initializer 로 처리하도록 수정

wongbingg committed:
refactor : TodoListViewModel수정 및 적용

- TodoListViewModel 의 클로저 데이터바인딩
- Coordinator에 있던 Navigation item 설정코드를 다시 ViewController로 이동( Coordinator가 뷰전환만 관리하도록 )

wongbingg committed:
refactor : 하위뷰인 ListVew, HeaderView 수정

- 불필요한 코드 제거
- initializer에서 받아오는 프로퍼티는 선언시 습관적으로 해주었던 옵셔널? 을 제거. (category, viewModel)
- 데이터 세팅로직 일부 수정

wongbingg committed:
refactor : ModalTransitionScene의 컨벤션 통일 및 코드수정

- 마크주석 분리
- 불필요한 옵셔널 선언 제거
- 일부 로직 수정

wongbingg committed:
refactor : PopoverViewController 컨벤션 통일

- private enum의 네이밍을 구체적으로 지정
- 마크주석으로 분리
- create 메서드를 이용한 생성에서 initializer를 이용한 생성으로 변경

wongbingg committed:
fix : formsheet 에서 키보드가 텍스트뷰를 가리는 오류 해결

wongbingg committed:
docs : ReadMe 업데이트

wongbingg committed:
refactor : viewModel 을 Input과 Output으로 분리

wongbingg committed:
refactor : viewController에 있던 바인딩을 각 뷰로 이동

- HeaderView에서 didChangedCount 클로저 바인딩
- ListCollectionView에서 didCreatedTodo, didEditedTodo, didMovedTodo 클로저 바인딩
- 디미터 법칙을 준수
- etc: final 처리
